File: frontend/src/api/patient.ts
/** 患者端 API 服务层 */

import { API_BASE } from './client';

import { authHeaders } from './client';

/** 后端 HealthProfileResponse 的实际返回结构 */
export interface BackendProfile {
  user_id: string;
  date_of_birth: string | null;
  gender: string | null;
  blood_type: string | null;
  height: number | null;
  weight: number | null;
  allergies: string[];
  chronic_diseases: string[];
  medications: Array<{
    name: string;
    dosage: string;
    frequency: string;
    start_date: string;
    end_date?: string | null;
  }>;
  created_at: string;
  updated_at: string;
}

export interface PatientProfile {
  id: string;
  name: string;
  email: string;
  phone?: string;
  date_of_birth?: string;
  gender?: string;
  blood_type?: string;
  height?: number;
  weight?: number;
  allergies?: string[];
  chronic_diseases?: string[];
  medications?: Array<{
    name: string;
    dosage: string;
    frequency: string;
    start_date?: string;
  }>;
}

export interface MedicalCase {
  id: string;
  title: string;
  description: string;
  status: string;
  created_at: string;
  diagnosis?: string;
}

export interface CarePlan {
  id: string;
  title: string;
  goals: string[];
  status: string;
  tasks: Array<{
    id: string;
    description: string;
    due_date?: string;
    status: string;
    order?: number;
  }>;
  start_date: string;
  end_date?: string;
}

export async function getProfile(): Promise<BackendProfile> {
  const res = await fetch(`${API_BASE}/patient/profile`, { headers: authHeaders() });
  if (!res.ok) throw new Error('Failed to fetch profile');
  return res.json();
}

export async function updateProfile(data: Partial<PatientProfile>): Promise<BackendProfile> {
  const res = await fetch(`${API_BASE}/patient/profile`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    let detail = `保存失败 (${res.status})`;
    try { const j = JSON.parse(body); detail = j.detail || detail; } catch {}
    throw new Error(detail);
  }
  return res.json();
}

export async function listCases(): Promise<MedicalCase[]> {
  const res = await fetch(`${API_BASE}/patient/cases`, { headers: authHeaders() });
  if (!res.ok) throw new Error('Failed to fetch cases');
  return res.json();
}

export async function listCarePlans(): Promise<CarePlan[]> {
  const res = await fetch(`${API_BASE}/patient/care-plans?include_tasks=true&size=50`, { headers: authHeaders() });
  if (!res.ok) throw new Error('Failed to fetch care plans');
  const data = await res.json();
  // Backend returns paginated response { items: [...], total, page, size, pages }
  return (data?.items ?? data) as CarePlan[];
}

export async function ackTask(planId: string, taskId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/patient/care-plans/${planId}/tasks/${taskId}/complete`, {
    method: 'POST',
    headers: authHeaders(),
  });
  if (!res.ok) throw new Error('Failed to complete task');
}

// ── Medication Reminders ──────────────────────────────────────────────

export interface MedicationReminder {
  id: string;
  name: string;
  dosage: string;
  frequency: string;
  time_slots: string[];
  start_date: string;
  end_date?: string;
  status: string;
  note?: string;
  today_records?: Array<{ id: string; scheduled_time: string; status: string; taken_at?: string }>;
  created_at: string;
}

export interface TodayMedicationItem {
  reminder_id: string;
  record_id: string;
  name: string;
  dosage: string;
  scheduled_time: string;
  taken_at?: string;
  status: string;
}

export interface TodayMedicationResponse {
  items: TodayMedicationItem[];
  taken_count: number;
  pending_count: number;
  total_count: number;
}

export async function createMedication(data: {
  name: string; dosage: string; frequency: string; time_slots?: string[];
  lead_minutes?: number; remind_enabled?: boolean;
}): Promise<MedicationReminder> {
  const res = await fetch(`${API_BASE}/patient/medications`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error('Failed to create medication');
  return res.json();
}

export async function listMedications(): Promise<MedicationReminder[]> {
  const res = await fetch(`${API_BASE}/patient/medications`, { headers: authHeaders() });
  if (!res.ok) throw new Error('Failed to fetch medications');
  return res.json();
}

export async function getTodayMedications(): Promise<TodayMedicationResponse> {
  const res = await fetch(`${API_BASE}/patient/medications/today`, { headers: authHeaders() });
  if (!res.ok) throw new Error('Failed to fetch today medications');
  return res.json();
}

export async function takeMedication(medicationId: string): Promise<{ id: string; status: string; taken_at?: string }> {
  const res = await fetch(`${API_BASE}/patient/medications/${medicationId}/take`, {
    method: 'POST',
    headers: authHeaders(),
  });
  if (!res.ok) throw new Error('Failed to take medication');
  return res.json();
}

export async function deleteMedication(medicationId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/patient/medications/${medicationId}`, {
    method: 'DELETE',
    headers: authHeaders(),
  });
  if (!res.ok) throw new Error('Failed to delete medication');
}

File: backend/app/schemas/patient.py
"""Patient-facing API schemas: health profile, care plans, care tasks."""

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.patient_profile import (
    BloodTypeEnum,
    GenderEnum,
    PlanStatusEnum,
    TaskStatusEnum,
)


# ── Health Profile ─────────────────────────────────────────────────────


class MedicationSchema(BaseModel):
    """A single medication entry in the patient's profile."""

    name: str = Field(..., min_length=1)
    dosage: str = Field(..., min_length=1)  # string like "5mg", "2片"
    frequency: str = Field(..., min_length=1)
    start_date: date
    end_date: date | None = None


class HealthProfileResponse(BaseModel):
    """Full health profile returned to the patient."""

    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    date_of_birth: date | None = None
    gender: GenderEnum | None = None
    blood_type: BloodTypeEnum | None = None
    height: Decimal | None = None
    weight: Decimal | None = None
    allergies: list[str] = []
    chronic_diseases: list[str] = []
    medications: list[MedicationSchema] = []
    created_at: datetime
    updated_at: datetime


class HealthProfileUpdate(BaseModel):
    """Partial update payload for health profile.

    Scalar fields are direct-replace. JSONB-list fields (allergies,
    chronic_diseases, medications) are full-replace — the client must
    send the complete new array.
    """

    date_of_birth: date | None = None
    gender: GenderEnum | None = None
    blood_type: BloodTypeEnum | None = None
    height: Decimal | None = Field(None, ge=30, le=300)
    weight: Decimal | None = Field(None, ge=2, le=500)
    allergies: list[str] | None = None
    chronic_diseases: list[str] | None = None
    medications: list[MedicationSchema] | None = None


class AllergyOperation(BaseModel):
    """Single allergy string for add/remove operations."""

    item: str = Field(..., min_length=1)


# ── Care Plan ──────────────────────────────────────────────────────────


class CareTaskResponse(BaseModel):
    """A single task within a care plan."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    description: str
    due_date: date
    status: TaskStatusEnum
    completed_at: datetime | None = None
    order: int


class CarePlanResponse(BaseModel):
    """Care plan returned to the patient."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    goals: list[str] = []
    status: PlanStatusEnum
    start_date: date
    end_date: date | None = None
    tasks: list[CareTaskResponse] | None = None
    created_at: datetime
    updated_at: datetime


class CarePlanListResponse(BaseModel):
    """Paginated list of care plans."""

    items: list[CarePlanResponse]
    total: int
    page: int
    size: int
    pages: int


class TaskListResponse(BaseModel):
    """Paginated list of tasks within a plan."""

    items: list[CareTaskResponse]
    total: int
    page: int
    size: int
    pages: int


class CareTaskActionResponse(BaseModel):
    """Response after completing or skipping a task."""

    id: uuid.UUID
    status: TaskStatusEnum
    completed_at: datetime | None = None


class CareTaskCreate(BaseModel):
    """Input for creating a single task within a care plan."""

    description: str = Field(..., min_length=1, max_length=500)
    due_date: date | None = None


class CarePlanCreate(BaseModel):
    """Input for creating a new care plan (defaults to active)."""

    title: str = Field(..., min_length=1, max_length=200)
    goals: list[str] = []
    start_date: date | None = None
    end_date: date | None = None
    tasks: list[CareTaskCreate] = []


# ── Medication Reminder ─────────────────────────────────────────────────


class MedicationReminderCreate(BaseModel):
    """Input for adding a medication reminder."""

    name: str = Field(..., min_length=1, max_length=200)
    dosage: str = Field(..., min_length=1, max_length=100)
    frequency: str = Field(..., min_length=1, max_length=100)
    time_slots: list[str] = Field(default_factory=lambda: ["08:00", "20:00"])
    start_date: date | None = None
    end_date: date | None = None
    note: str | None = Field(None, max_length=500)
    lead_minutes: int = Field(default=15, ge=1, le=1440)
    remind_enabled: bool = True


class MedicationReminderUpdate(BaseModel):
    """Input for updating a medication reminder."""

    name: str | None = None
    dosage: str | None = None
    frequency: str | None = None
    time_slots: list[str] | None = None
    start_date: date | None = None
    end_date: date | None = None
    note: str | None = None
    lead_minutes: int | None = Field(None, ge=1, le=1440)
    remind_enabled: bool | None = None


class MedicationRecordResponse(BaseModel):
    """A single dose record."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    scheduled_date: date
    scheduled_time: str
    taken_at: datetime | None = None
    status: TaskStatusEnum


class MedicationReminderResponse(BaseModel):
    """Medication reminder with today's records."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    dosage: str
    frequency: str
    time_slots: list[str] = []
    start_date: date
    end_date: date | None = None
    status: PlanStatusEnum
    note: str | None = None
    lead_minutes: int = 15
    remind_enabled: bool = True
    last_reminded_at: datetime | None = None
    reminded_count: int = 0
    today_records: list[MedicationRecordResponse] = []
    created_at: datetime
    updated_at: datetime


class TodayMedicationItem(BaseModel):
    """A single pending/taken dose for today's view."""

    reminder_id: uuid.UUID
    record_id: uuid.UUID
    name: str
    dosage: str
    scheduled_time: str
    taken_at: datetime | None = None
    status: TaskStatusEnum


class TodayMedicationResponse(BaseModel):
    """Today's medication overview."""

    items: list[TodayMedicationItem]
    taken_count: int
    pending_count: int
    total_count: int


class PlanStatusUpdate(BaseModel):
    """Request to update plan status."""

    status: PlanStatusEnum


# ── Cases (thin proxy for MedicalCase) ─────────────────────────────────


class CaseListResponseItem(BaseModel):
    """Patient-visible fields from MedicalCase."""

    id: uuid.UUID
    chief_complaint: str | None = None
    ai_diagnosis_summary: str | None = None
    severity: str | None = None
    is_emergency: bool = False
    status: str
    created_at: datetime
    updated_at: datetime


class CaseListResponse(BaseModel):
    """Paginated list of patient cases."""

    items: list[CaseListResponseItem]
    total: int
    page: int
    size: int
    pages: int


# ── Generic pagination params (used by routes) ─────────────────────────


class PaginationParams:
    """Pydantic model for query-string pagination."""

    def __init__(self, page: int = 1, size: int = 20):
        self.page = max(1, page)
        self.size = min(100, max(1, size))

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.size

    @property
    def limit(self) -> int:
        return self.size
